WIP: Adds Ginkgo interface

MarcelKoch opened the pull request:
This is a draft for the Ginkgo interface as discussed in #15190. As suggested in the discussion, instead of adding new wrapper classes for Ginkgo objects, this interface allows for easy conversion between deal.II and Ginkgo objects. 

Currently it implements the following:
- create a Ginkgo vector (dense matrix) from a deal.II array view. ~~This comes in a two different flavors: `create_vector(ArrayView)` creates a view, while `create_vector(Executor, ArrayView)` will create a copy if the Ginkgo Executor doesn't match the memory space of the array view (otherwise it also creates a view).~~ It is always required to pass-in a Ginkgo `Executor` that matches the `MemorySpace` used by the array view. If the executor doesn't match, an exception will be thrown.
- create a Ginkgo Csr matrix from a deal.II SparseMatrix. This always copies the data. Other Ginkgo sparse matrix formats can be added easily.
- create a deal.II LinearOperator from a Ginkgo LinOp. The domain and range is fixed to `LinearAlgebra::distributed::Vector`, since that is the only native deal.II vector type that supports device memory.

I've also added a fix that preserves the C++ compiler for testing. This was necessary since I ran these tests on a HIP system, which requires using the hip compiler as C++ compiler.

MarcelKoch committed:
preserve cxx compiler for tests

MarcelKoch committed:
adds tests for ginkgo interface

MarcelKoch committed:
adds ginkgo vector interface

MarcelKoch committed:
adds linear operator interface

MarcelKoch committed:
adds ginkgo matrix interface

MarcelKoch commented:
Pinging @drwells that our discussion form #15190 can be continued here.

drwells commented:
Thanks for posting this - I've been excessively busy recently but I'll be back-to-normal on Monday. I'll take a look then.

bangerth commented:
@MarcelKoch Do I understand correctly that this is the current patch? I know we've been sitting on this for a long time now -- let's see whether we can revisit this now. Would you mind rebasing the patch?

MarcelKoch committed:
make executor function argument

MarcelKoch committed:
update license

MarcelKoch commented:
@bangerth Thanks for checking in. I've rebased the branch and added some smaller changes. I think for now this is a minimal working interface.

There are still some open questions:
- How should the different matrix formats in Ginkgo be exposed? Right now, I would favor functions `create_<fmt>_matrix` like the already existing one.
- Right now this only supports sequential deal.ii types. I think that is reasonable, but it might still make sense to discuss how distributed vectors/matrices could be supported. 
 (Distributed functionality could later added into a `GinkgoInterface::MPI` namespace, so that it won't break the existing interface.) 

Todos:
- [ ] Add documentation

MarcelKoch committed:
fixup! make executor function argument

bangerth commented:
We dropped the ball here :-( Let me postpone this until after the release, but let me also ask: Are you still interested in moving forward with the Gingko interfaces? If so, let's make sure we get this patch reviewed in the not too far future.

MarcelKoch commented:
Thanks for the update. I'm still interested in moving forward with this. I will update this branch to make it compatible with the latest master branch.

bangerth commented:
Sounds good. Please ping us when you're at a point for someone to take a look!
